cspyce/spice_cell.py

```diff
@@ -0,0 +1,170 @@
+            array_descriptor = np.dtype(np.double)
+        elif typeno == SPICE_CELL_CHAR:
+            array_descriptor = np.dtype(("S", length))
+            length = max(2, length)
+        else:
+            raise ValueError(f"Bad type {typeno} passed to SpiceCell init")
+
+        if data is not None:
+            data = np.asarray(data, dtype=array_descriptor)
+            size = max(size, len(data) + 6)  # add a little bit of spare room
+
+        self._header = np.rec.array(None, SPICE_CELL_HEADER_DESCRIPTOR, 1)[0]
+        self._descriptor = array_descriptor
+
+        self._header.dtype_ = typeno
+        self._header.length_ = length
+        self._header.size_ = size
+        self._header.card_ = 0
+        self._header.isSet_ = True
+        self._header.adjust_ = False
+        self._header.init_ = False
+
+        self.__grow_array(size, init=True)
+
+        if data is not None:
+            self._user_data[:len(data)] = data
+            self._header.card_ = len(data)
+
+    def __getitem__(self, index):
+        if -self.card <= index < 0:
+            index += self.card
+        return self._user_data[index]
+
+    def __setitem__(self, index, value):
+        if -self.card <= index < 0:
+            index += self.card
+        self._user_data[index] = value
+
+    def __len__(self):
+        return self._header.card_.item()
+
+    def __iter__(self):
+        for i in range(self.card):
+            yield self[i]
+
+    def clear(self):
+        self.card = 0
+
+    def append(self, value):
+        if self.card == self.size:
+            self.__grow_array(max(10, 2 * self.size))
+        self[self.card] = value
+        self.card += 1
+
+    def extend(self, values):
+        values = np.asarray(values, self._descriptor)
+        count = len(values)
+        if self.card + count > self.size:
+            self.__grow_array(max(self.card + count + 10, 2 * self.size))
+        self._user_data[self.card : self.card + count] = values
+        self.card += count
+
+    def __iadd__(self, values):
+        self.extend(values)
+        return self
+
+    @property
+    def size(self):
+        return self._header.size_.item()
+
+    @size.setter
+    def size(self, size):
+        assert isinstance(size, int) and size > 0
+        self.__grow_array(size)
+
+    @property
+    def card(self):
+        return len(self)
+
+    @card.setter
+    def card(self, value):
+        assert isinstance(value, int) and 0 <= value <= self.size
+        self._header.card_ = value
+
+    def as_array(self):
+        return self._user_data[0:self.card]
+
+    def as_intervals(self):
+        return self.as_array().reshape(-1, 2)
+
+    def __str__(self):
+        typename = ['char', 'float', 'int'][self._header.dtype_]
+        return f"<SpiceCell {typename} {self._header.card_}/{self._header.size_} {self.as_array()}>"
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __grow_array(self, size, init=False):
```

Comment:
And why does this have double underscores? That's usually reserved for system-standard functions.

Reply:
Actually no.  Double underscore at both the beginning and the ending are system internal.  Using __ as prefix but not a suffix causes name mangling.  The name of the method isn't actually __grow_array, but something like _SpiceCell__grow_array.  At Google, this was used as a strong hint that this is an extremely internal function and that outsiders shouldn't touch. This is Python, so this is the closet you can get to a private method.